### PR TITLE
Supports __future__.annotations

### LINKED ISSUE #7907: Implement PEP 585 (Generic builtins and __future__.annotations)
* Are you reporting a bug, or opening a feature request? Reporting a bug
* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate if you try to simplify your case to a minimal repro.
```py
from __future__ import annotations
from collections import defaultdict

squares: defaultdict[tuple, int] = defaultdict(lambda: 999999)
```
* What is the actual behaviour/output? MyPy tells me `"defaultdict" is not subscriptable, use "typing.DefaultDict" instead`
* What is the behaviour/output you expect? MyPy ignores the subscript error (and recognises it as equivalent) because of `from __future__ import annotations`
* What is the version of mypy and Python you are using? 0.740 on 3.8.0 (Ubuntu 18.04)
* No additional MyPy flags

### PR DESCRIPTION
**Pull request by TH3CHARLie**
This PR partially addresses #7907.

### CODE REVIEW

**Comment by TH3CHARLie** [comment]
However, tests are lacked. I wonder in which unit-test file should the test be implemented. @ilevkivskyi

**Comment by JukkaL** [comment]
You can add a new test file, since no existing file seems to be a very good fit. Maybe `test-data/unit/check-future.test`?

**Comment by TH3CHARLie** [comment]
> You can add a new test file, since no existing file seems to be a very good fit. Maybe `test-data/unit/check-future.test`?

Good idea, I will get it done once I wake up tomorrow

**Comment by TH3CHARLie** [comment]
Note: after adding `defaultdict` stub to `test-data/unit/lib-stub/collections.pyi`, my vscode's pylint plugin telling me `class already defined line 2pylint(function-redefined)`

File: test-data/unit/lib-stub/collections.pyi [hunk 1]
```diff
@@ -9,3 +9,50 @@ def namedtuple(
...
+    Sized as Sized,
+    ValuesView as ValuesView,
+)
+
+_S = TypeVar('_S')
+_T = TypeVar('_T')
+_KT = TypeVar('_KT')
+_VT = TypeVar('_VT')
+
+class defaultdict(Dict[_KT, _VT], Generic[_KT, _VT]):
```
[the remark this answers is not in the document]
Comment (TH3CHARLie) [reply test-data/unit/lib-stub/collections.pyi]: Agreed. I'll fix it once along with further changes if we have an agreement in #7907

**Comment by ethanhs** [comment]
@TH3CHARLie would you like to resurrect this PR?

I believe the semantics are better understood now.

**Comment by TH3CHARLie** [comment]
> @TH3CHARLie would you like to resurrect this PR?
> 
> I believe the semantics are better defined now.

Great, I'll read through the new PEP and update the changes in this PR.

**Comment by ethanhs** [comment]
@TH3CHARLie note that the total set of things being supported is more than just those enumerated in the PEP.

**Comment by TH3CHARLie** [comment]
@ethanhs If I understand correctly, to support PEP585 in mypy, all items listed in https://github.com/gvanrossum/cpython/pull/1#issuecomment-582781121 can now be parameterized, so this PR actually needs to spot such use cases and suppress currently reported errors, right?

**Comment by TH3CHARLie** [comment]
@ethanhs If I understand correctly, to support PEP585 in mypy, all items listed in https://github.com/gvanrossum/cpython/pull/1#issuecomment-582781121 can now be parameterized, so this PR actually needs to spot such use cases and suppress currently reported errors, right?

**Comment by ethanhs** [comment]
@TH3CHARLie that is essentially correct. I think the best way to think of it is "anything in the stdlib that is Generic in typeshed". That may be easier to implement too.

**Comment by TH3CHARLie** [comment]
> anything in the stdlib that is Generic in typeshed

I had one initial version now but it's simply enumerating all the cases using a long list so I am interested in you mentioned "anything in the stdlib that is Generic in typeshed". Can this be achieved through some interaction with typeshed so mypy computes the generic list automatically?(I'm actually not pretty familiar with typeshed codebase)

**Comment by TH3CHARLie** [comment]
Sorry for the delay. I carelessly forget to push my initial version but thanks to this I think twice and find my first attempt to suppress more errors than it should.

The PR now suppress errors on generic built-ins, the only thing tricky is to handle `type`. @ethanhs

**Comment by TH3CHARLie** [comment]
ping @JukkaL  and @ethanhs  for review

**Comment by AllanDaemon** [comment]
Is this planned to be released before Python 3.9 release, that will implement PEP 585?

**Comment by TH3CHARLie** [comment]
> Is this planned to be released before Python 3.9 release, that will implement PEP 585?

Thanks for asking!  This PR is delayed since we spent a significant amount of time working on compiler kinds of stuff. I'll try to mention this with core developers and move this forward.

**Comment by ethanhs** [comment]
@TH3CHARLie I've read over the PR, but I believe as written it currently only supports `__future__.annotations`, not the full PEP 585, which adds support for sub-scripting builtins on 3.9+ without the future import. I think it might be better to land this and implement PEP 585 in another PR.

**Review by ethanhs — APPROVED** [review]
Thank you for working on this!

**Comment by TH3CHARLie** [comment]
> @TH3CHARLie I've read over the PR, but I believe as written it currently only supports `__future__.annotations`, not the full PEP 585, which adds support for sub-scripting builtins on 3.9+ without the future import. I think it might be better to land this and implement PEP 585 in another PR.

Thanks for the review! then I'll rename this PR and merge it.

**Pull request merged**